Add C++ bindings for listen points.

zacharygrafton opened the pull request.

davidmoreno commented on src/bindings/cpp/onion.hpp:
```diff
@@ -26,6 +26,7 @@
 
 #include <string>
 #include <onion/onion.h>
+#include "listen_point.hpp"
```
As it is using LitenPoint& there is no need to fully include listen_point. It is enough with a forward class declaration "class ListenPoint;" (inside the namespace). This speeds up compilation times.

zacharygrafton replied:
When I remove the include and replace it with a forward declaration, I get an error with the build. I believe that the full include is needed because in the addListenPoint method, I tried to keep the code inline and it needs access to the c_handler method. I don't think there is anyway around it.

davidmoreno commented:
Everything looks OK, but I think an example or test should be included, to ensure it all works as expected and there are no memory leaks.

If you can manage to include one, great, if not I will delay the merge until I can find time to create one.

Regards, David.

zacharygrafton commented:
David,
I'm actually 100% sure it leaks memory if the ListenPoint is never added to an Onion object, however, if it is added to an Onion object, the listen point is freed when the server shuts down. I have actually tried approaching this from multiple angles, including one where the user could actually derive and add functionality to the listen points(however, that led to a decent bit of code duplication.) I'm not really sure what we should expect from the user in this case. It would seem mostly pointless to create a ListenPoint without adding it to an Onion object, so I took the path that assumed this would be the case. I'll add an example and document the expected usage.

Thanks,
Zack

zacharygrafton commented:
David,

I added an example that reproduces the current multiport example, but using the C++ bindings instead. I added some doxygen documentation on the correct usage as well.

Zack
